nixos/modules/services/matrix/draupnir.nix

```diff
@@ -0,0 +1,305 @@
+            description = ''
+              Account name on the Matrix homeserver.
+            '';
+          };
+
+          passwordFile = lib.mkOption {
+            type = with lib.types; nullOr path;
+            default = null;
+            description = ''
+              File containing the password for the Matrix account.
+              Make sure this does not contain newlines if writing manually: `:set noeol nofixeol` for vim or -L for nano.
+            '';
+          };
+
+          options = lib.mkOption {
+            type = lib.types.submodule (import ./pantalaimon-options.nix { inherit lib; inherit config; name = "draupnir"; });
+            default = { };
+            description = ''
+              Pass through additional options to the `pantalaimon` service.
+            '';
+          };
+        };
+      };
+    };
+    #endregion
+
+    #region Draupnir settings
+    settings = lib.mkOption {
+      example = lib.literalExpression ''
+        {
+          autojoinOnlyIfManager = true;
+          automaticallyRedactForReasons = [ "spam" "advertising" ];
+        }
+      '';
+      description = ''
+        Draupnir settings (see [Draupnir's default configuration](https://github.com/the-draupnir-project/Draupnir/blob/main/config/default.yaml) for available settings).
+        These settings will override settings made by the module config.
+      '';
+      default = { };
+      type = lib.types.submodule {
+        freeformType = format.type;
+        options = {
+          #region Readonly settings - these settings are not configurable
+          dataPath = lib.mkOption {
+            type = lib.types.str;
+            default = "/var/lib/draupnir";
+            readOnly = true;
+            description = ''
+              The path where Draupnir stores its data.
+
+              ::: {.note}
+                If you want to customize where this data is stored, use a bind mount.
+              :::
+            '';
+          };
+
+          pantalaimon = lib.mkOption {
+            readOnly = true;
+            description = ''
+              `pantalaimon` settings (enables E2E Encryption support).
+              This property is read-only, please configure `services.draupnir.pantalaimon` instead!
+            '';
+            default = {};
+            type = lib.types.submodule {
+              freeformType = format.type;
+              options = {
+                use = lib.mkOption {
+                  type = lib.types.bool;
+                  default = cfg.pantalaimon.enable;
+                  defaultText = "cfg.pantalaimon.enable";
+                  readOnly = true;
+                  description = ''
+                    Whether to use `pantalaimon` for E2E encryption. Enabled if `services.draupnir.pantalaimon.enable` is `true`.
+                  '';
+                };
+                username = lib.mkOption {
+                  type = lib.types.str;
+                  default = cfg.pantalaimon.username;
```

Comment:
I honestly don't even remember what this entire readonly section is for...

Reply:
Did some digging, I remember what it's for: that's the part that ends up in the Draupnir config file... I think it might be better to drop that section and define `services.draupnir.settings.pantalaimon` in the `config` section?

Reply:
This is what I ended up doing, and it seems to work as intended while involving a lot less code :)